sdk/inc/azure/core/az_mqtt5.h

```diff
@@ -0,0 +1,784 @@
+
+  /**
+   * @brief The payload of the message.
+   */
+  az_span payload;
+
+  /**
+   * @brief The quality of service of the message.
+   */
+  int8_t qos;
+
+  /**
+   * @brief The ID of the message.
+   */
+  int32_t id;
+} az_mqtt5_recv_data;
+
+/**
+ * @brief MQTT 5 publish acknowledgement data.
+ *
+ */
+typedef struct
+{
+  /**
+   * @brief The publish request ID.
+   */
+  int32_t id;
+} az_mqtt5_puback_data;
+
+/**
+ * @brief MQTT 5 subscribe data.
+ *
+ */
+typedef struct
+{
+  /**
+   * @brief The properties of subscribe request.
+   */
+  az_mqtt5_property_bag* properties;
+  /**
+   * @brief Topic filter to subscribe to.
+   */
+  az_span topic_filter;
+
+  /**
+   * @brief Quality of service of the subscription.
+   */
+  int8_t qos;
+
+  /**
+   * @brief Id to correlate the subscription request with the response acknowledgement.
+   */
+  int32_t out_id;
+} az_mqtt5_sub_data;
+
+/**
+ * @brief MQTT 5 subscribe acknowledgement data.
+ *
+ */
+typedef struct
+{
+  /**
+   * @brief The subscribe request ID.
+   */
+  int32_t id;
+} az_mqtt5_suback_data;
+
+/**
+ * @brief MQTT 5 connect data.
+ *
+ */
+typedef struct
+{
+  /**
+   * @brief The properties of the connect request.
+   */
+  az_mqtt5_property_bag* properties;
+
+  /**
+   * @brief Hostname or IP address of the MQTT 5 broker.
+   */
+  az_span host;
+
+  /**
+   * @brief Port number broker is listening on.
+   */
+  int16_t port;
+
+  /**
+   * @brief The username to send to the broker.
+   */
+  az_span username;
+
+  /**
+   * @brief The password to send to the broker.
+   */
+  az_span password;
+
+  /**
+   * @brief The client ID to send to the broker.
+   */
+  az_span client_id;
+
+  /**
+   * @brief The certificate to use for authentication.
+   */
+  az_mqtt5_x509_client_certificate certificate;
+} az_mqtt5_connect_data;
+
+/**
+ * @brief MQTT 5 connect acknowledgement data.
+ *
+ */
+typedef struct
+{
+  /**
+   * @brief Connection acknowledgement reason code. Indicates success or reason for failure.
+   */
+  int32_t connack_reason;
+
+  /**
+   * @brief Indicates whether a TLS authentication error occurred.
+   */
+  bool tls_authentication_error;
+} az_mqtt5_connack_data;
+
+/**
+ * @brief MQTT 5 disconnect data.
+ *
+ */
+typedef struct
+{
+  /**
+   * @brief Indicates whether a TLS authentication error occurred.
+   */
+  bool tls_authentication_error;
+
+  /**
+   * @brief Indicates whether the disconnect was requested by the client.
+   */
+  bool disconnect_requested;
+} az_mqtt5_disconnect_data;
+
+/**
+ * @brief MQTT 5 property string.
+ */
+typedef struct
+{
+  /**
+   * @brief The string value of the property.
+   */
+  az_span str;
+} az_mqtt5_property_string;
+
+/**
+ * @brief Gets the string value of an MQTT 5 property string.
+ *
+ * @param[in] prop_str The MQTT 5 property string.
+ *
+ * @return The string value of the property.
+ */
+AZ_NODISCARD AZ_INLINE az_span az_mqtt5_property_string_get(az_mqtt5_property_string* prop_str)
```

Comment:
Correct - str should be under `_internal`

Reply:
The type az_mqtt5_property_string should be different for different MQTT implementations.

E.g. for ESP-MQTT it won't need two `az_span`s: instead, it will call into `esp_mqtt5_client_get_user_property` ([example](https://github.com/espressif/esp-idf/blob/28167ea5a3cdd49227216c4131a9df7f4eda260e/examples/protocols/mqtt5/main/app_main.c#L77C17-L77C51)).

The struct in that case would need to hold on to the key/value pointers in order to allow freeing them later.

Reply:
Removed struct declaration for all `az_mqtt5_property_*`. They are now declared inside `az_mqtt5_[platform implementation]`. Removed inline attribute of the getter functions and they are also now declared in `az_mqtt5_[platform implementation]`.